### ISSUE TITLE
Reduce reliance on docker containers for module tests

### ISSUE THREAD

**Issue by stefpiatek** [UCLH-Foundry/PIXL#237]
### Definition of Done / Acceptance Criteria

Use mocking and monkeypatching to reduce the need to spin up docker containers for modules

### Details and Comments

`pixl_core` will need to have the FTPS server to test that it can write to the server, but otherwise investigate which services are not needed

**Comment by jeremyestein** [comment]
For the sake of those containers that are still needed for testing, we could look into speeding up the image builds.
The usual trick of installing your requirements separately doesn't work for us because we're using pyproject.toml files:
```
COPY requirements.txt .
RUN pip install -r requirements.txt
COPY . .
RUN pip install
```

So we end up with some very long running `pip install` commands that download all the dependencies and re-install them when only our code has changed.

**Comment by jstutters** [comment]
[pip-tools](https://github.com/jazzband/pip-tools) can give you `requirements.txt` from a `pyproject.toml` but then you have the headache of keeping those in sync.  Run the tests from a Makefile?
```
python -m piptools compile -o requirements.txt pyproject.toml
```

**Comment by stefpiatek** [comment]
> [pip-tools](https://github.com/jazzband/pip-tools) can give you requirements.txt from a pyproject.toml but then you have the headache of keeping those in sync. Run the tests from a Makefile?

Yeah as this is all to go with steps in dockerfiles, can just have that as an earlier step and then install the dependencies before we install the package

**Comment by peshence** [comment]
Me and @ruaridhg had a look at the CLI module, where the only tests needing the rabbitmq docker are in [test_queue_start_and_stop_parquet.py](https://github.com/UCLH-Foundry/PIXL/blob/main/cli/tests/test_queue_start_and_stop_parquet.py). We started to mock out the consumer and producer but found ourselves wondering if we're actually testing anything then? I would argue all the tests there are unnecessary

**Comment by peshence** [comment]
Looking for input, details below, but basic questions are:
- do we value functional tests of each module? e.g. `pixl_core/tests/patient_queue/{producer/subscriber}.py` assume yes as there is little logic in the underlying code but we need to have some sort of handle on its condition.
- In order to test a lot of our code we need running services. Instead of bringing up everything each time, we could simply run within a dev container with all the services running. This would present some load onto developers' systems. Is everybody up for it? Do you have alternative suggestions?

## CLI
only needs a rabbitmq instance for the `cli/tests/test_queue_start_and_stop_parquet.py` tests which try to:
- populate - does tons of stuff
	- [ ] **raise ticket - exhausts state file but ignores new directory - prioritise [#198](https://github.com/UCLH-Foundry/PIXL/issues/198) to remove state files and fix**
	- [ ] **Mock producer**
	- [ ] **mock statefile loading or remove as per #198**
- empty queue 
	- also populates
	- stops queues and saves to csv (for some reason not a temporary one and dependent on folder you are running from) - this is effectively a wrapper of `PIXLBlockingConsumer.consume_all()`
	- [ ] **DELETE**
- check queue is up - literally that, tests underlying framework/integration 
    - [ ] **DELETE**
## Core
### FTP-Server
tests FTPS upload. Ticket says we should expect to keep this but I don’t see why we couldn’t mock it or run an [ftp server from python](https://pytest-localftpserver.readthedocs.io/en/latest/usage.html)? @stef @milan
### Patient Queue
Here we test basic implementations of a pika-based AMQP producer and consumer (async and blocking) and we use docker.
- `test_producer` simply check if connection can be made, queue can be cleared, message can be published → trust underlying library and integration/system tests 
    - [ ] **can also add an assert in test_consume_all to replace this whole thing** 
    - [ ] **DELETE**
- `test_subscriber` - both tests effectively test producer as well.
	- async `test_create` should just check that a callback is called upon finding a message - again unnecessary. Also is not actually awaited and therefore always passes. Adding `pytest.mark.asyncio` and removing weird event loop thing does not fix it 
	    - [ ] create an async test that runs
	- [ ] `test_consume_all` add a check that file is created; use pytest tmp fixture to delete file automatically

#### Queue wrappers testing options
- try to mock pika (underlying rabbitmq client framework)
	- a lot of work (?)
	- reduces the usefulness of test
- try to mock AMQP server e.g. with [amqp-mock](https://pypi.org/project/amqp-mock/)
	- unknown amount of work
	- introducing potentially limited server into testing
- run locally installed rabbitmq (see pytest_rabbitmq)
	- requires you to have installed rabbitmq yourself
	- needs you to point to your installation of rabbitmq - varies from system to system, or at least between OS-s
	- could potentially do that with a devcontainer that has rabbitmq installed
- devcontainer
	- with a running rabbitmq docker container
	- or with an installed rabbitmq
	- If we include other modules' tests we will need
	    - ftps server (though could be abstracted imo)
	    - orthanc raw
	    - vna fake server
	    - ehr dbs - these will likely go away?
            - star db
            - postgres db
## PACS
Huge docker compose:
- pacsapi - a custom python docker image
- vna-qr - an orthanc image **NEEDS TO HAVE PLATFORM ADDED TO RUN ON MAC**
- ortanc raw - another orthanc image with some custom stuff
- queue - rabbitmq 

Tasks:
- [ ] convert so that test runs locally. 
    - Will need to reconfigure the Orthanc class to take in orthanc instance name and port from e.g. a config. Currently port gets resolved by docker compose, but host name and port are hardcoded and so can’t change (from `"http://orthanc-raw:8042"` to `"http://localhost:8044"`, untested)
- [ ] either
    - [ ] run docker compose from python
    - [ ] dev container
        - [ ] add vna and orthanc-raw to dev container (if we're doing that?)
- [ ] either mock the consumer and producer (isolate) or use rabbitmq from dev container (integrate)
- [ ] remove dangling orthanc test docker file

## SYSTEM
- runs for 6 minutes 
    - [ ] speed up with pip caches?
    - [ ] potentially speed up within the dev container? seems overly complex/not worth it
- creates a dangling `test.dcm` file 
    - [ ] **Fix**

**Comment by stef** [comment]
you wanted to hilite someone else...

**Comment by stefpiatek** [comment]
I _think_ we've effectively reduced the docker container usage in tests. If there's child tickets that you think should be made from this one then please make them, going to close!
